src/App.tsx

```diff
@@ -1,24 +1,186 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
-import { UserWarning } from './UserWarning';
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import { deleteTodoOnServer, getTodos, updateTodoOnServer } from './api/todos';
+import { filterTodos } from './helpers/filterTodos';
 
-const USER_ID = 0;
+import { TodoHeader } from './components/TodoHeader';
+import { TodoList } from './components/TodoList';
+import { TodoFooter } from './components/TodoFooter';
+import { ErrorNotification } from './components/ErrorNotification';
+
+import { Todo } from './types/Todo';
+
+import { FilterType } from './enums/FilterType';
+import { ErrorType } from './enums/ErrorType';
+import { UserIdContext } from './contexts/UserIdContext';
+import { OnChangeFunc } from './types/OnChangeFunc';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedFilter, setSelectedFilter] = useState(FilterType.All);
+  const [errorType, setErrorType] = useState(ErrorType.None);
+  const [isErrorShown, setIsErrorShown] = useState(false);
+  const [tempTodoTitle, setTempTodoTitle] = useState('');
+  const [isAllToggled, setIsAllToggled] = useState(false);
+  const [isClearCompleted, setIsClearCompleted] = useState(false);
+
+  const userId = useContext(UserIdContext);
+
+  useEffect(() => {
+    getTodos(userId)
+      .then((userTodos) => setTodos(userTodos))
+      .catch(() => {
+        setErrorType(ErrorType.Download);
+        setIsErrorShown(true);
+      });
+  }, []);
+
+  const activeTodosNum = useMemo(() => {
+    return todos.reduce((num, todo) => {
+      return todo.completed ? num : num + 1;
+    }, 0);
+  }, [todos]);
+  const completedTodosNum = todos.length - activeTodosNum;
+
+  const filteredTodos = useMemo(
+    () => filterTodos(todos, selectedFilter),
+    [selectedFilter, todos],
+  );
+
+  const showError = useCallback((error: ErrorType) => {
+    setErrorType(error);
+    setIsErrorShown(true);
+  }, []);
+  const hideError = useCallback(() => {
+    setIsErrorShown(false);
+  }, []);
+
+  const onAddTodo = useCallback(
+    (newTodo: Todo): void => setTodos((oldTodos) => [...oldTodos, newTodo]),
+    [],
+  );
+
+  const onDeleteTodo = useCallback((todoId: number): void => {
+    return setTodos((oldTodos) => {
+      return oldTodos.filter((todo) => todo.id !== todoId);
+    });
+  }, []);
```

Comment:
It's good that you use memoization 🔥  But memoization is a costly thing for memory. You don't need to fix it here if you don't want to. But if you’re interested in it, you can read more about it [here](https://kentcdodds.com/blog/usememo-and-usecallback), for example

Reply:
ok, thanks